documentation/architecture/decisions/0003-crate-feature-flags.md

```diff
@@ -1,24 +1,62 @@
+    - More broadly, it does not require the existence of an OS, and has no access to things like files or threads.
+
+2. `liballoc` (e.g. `alloc::*`): This is a superset of `libcore` which is allowed to perform dynamic memory allocation from a global memory allocator.
+    - The allocator may be user-provided (via the `#[global_allocator]` attribute), rather than the one provided by an OS or `libc`.
+    - This mostly contains collection types, such as `Vec`, `String`, `BTreeMap`, etc.
+    - Beyond the ability to perform global memory allocation, it has the same restrictions as `libcore`:
+        - No dependencies on system libraries.
+        - No assumptions about target environment beyond what is guaranteed by the target architecture.
+
+3. `libstd`, (e.g. `std::*`): This depends on (and is a superset of) both `liballoc` and `libcore`. It is what is used by default.
+    - Broadly speaking, it requires an OS that supports files and threads.
+        - There are exceptions here, in that a few targets exist which have `libstd` support but just return errors when this functionality is used.
+    - It can assume whatever it needs to about the target environment.
+        - It's allowed to contain `#[cfg(target_os = "...")]` statements internally.
+        - It's allowed to link against system-specific libraries on the OS (`libc`, `libm`, `libSystem.dylib`, `kernel32.dll`, ...).
+        - It can even place requirements on the OS version in use (to forbid Windows XP, for example)
+    - As it depends on `liballoc`, `libstd` is allowed to perform memory allocation anywhere it wants.
+        - If a `#[global_allocator]` is configured, it will do so out of that allocator
+        - If no `#[global_allocator]` is configured, then `libstd` will provide one based on the OS's default allocator.
+            - In the rare cases where `libstd` supports a target without a system-provided allocator, `libstd` will provide a default written in Rust (currently, it uses a Rust [port of `dlmalloc`](https://crates.io/crates/dlmalloc) for these situations).
+
+As you may note, these are analagous to the three configurations we wish to support, but there is an important difference:
+
+We should do not adopt the restriction that `libcore` and `liballoc` has around target-specific checks and `cfg`s and such. We favor practicality over purity, and without this, we would be completely unable to provide useful functionality for embedded environments.
 
 ## Decision
 
+To reflect these three scenarios, we will use three cargo features:
+1. `feature = "std"`, which allows full use of `libstd`. Mirroring the Rust stdlib, it implies `feature = "alloc"`.
+2. `feature = "alloc"`, which allows use of an allocator. As  `feature = "std"`.
+3. `feature = "no_std"` is used to add dependencies which are needed only in `no_std` configurations.
```

Comment:
This (`feature = "no_std"`) isn't common in the ecosystem from what I've seen, but it's also hard to imagine avoiding it.

That said, in general it's consider very bad to have two dependencies that error when they're both on, but it's not considered nearly as bad to have a feature you must turn on if another feature is disabled. This is largely because the former breaks library users and the latter does not.